File: Longest_Substring_Without_Repeating_Characters.py
'''
Created on 2014-8-6

Leetcode : Longest Substring Without Repeating Characters

Given a string, find the length of the longest substring without repeating 

characters. For example, the longest substring without repeating letters for 
"abcabcbb" is "abc", which the length is 3. For "bbbbb" the longest substring 
is "b", with the length of 1.



'''

class Solution:
    # @return an integer
    def lengthOfLongestSubstring(self, s):
        if len(s) == 0 :
            return 0
        maxLength = 1
        
        bottom = 0
        
        arr = [-1 for i in range(128)]
        
        for i in range(len(s)):
            pos = ord(s[i])
            
            if arr[pos] == -1:
                arr[pos] = i
              
            elif arr[pos] < bottom:
                    arr[pos] = i
                 
            else:
                maxLength = max(maxLength , i  - bottom)
                
                if arr[pos] >= bottom and i - arr[pos]  > maxLength:
                    bottom = arr[pos] + 1
                    maxLength = i - arr[pos] 
                   
                    arr[pos] = i
                else:
                    bottom = arr[pos] + 1
                    arr[pos] = i
                
        maxLength = max(len(s)  - bottom , maxLength)      
           
            
            
        return maxLength
test = Solution()
s='wlrbbmqbhcdarzowkkyhiddqscdxrjmowfrxsjybldbefsarcbynecdyggxxpklorellnmpapqfwkhopkmco'
#s = "qopubjguxhxdipfzwswybgfylqvjzhar"
print test.lengthOfLongestSubstring(s)
            
        

File: Cplusplus/Longest_Substring_Without_Repeating_Characters.cpp
/*
Leetcode : Longest_Substring_Without_Repeating_Characters.cpp

Given a string, find the length of the longest substring without

repeating characters. For example, the longest substring without

repeating letters for "abcabcbb" is "abc", which the length is 3.

For "bbbbb" the longest substring is "b", with the length of 1.
*/

#include <iostream>
#include <string>
#include <map>
using namespace std;



class Solution {
public:
    int lengthOfLongestSubstring(string s) {
            int maxL = 0;
            map<char , int> table;
            int pre = 0;
            map<char , int>::iterator iter;
            for(int i = 0 ; i < s.size() ; i++)
            {
                iter = table.find(s[i]);
                if(iter == table.end())
                {
                    table[s[i]] = i;
                }
                else
                {
                    if(table[s[i]] < pre)
                    {
                        table[s[i]] = i;
                    }
                    else
                    {
                        int len = i - pre;
                        maxL = maxL > len ? maxL:len;
                        pre = table[s[i]] + 1;
                        table[s[i]] = i;
                    }

                }
            }
            maxL = maxL > (s.size() - pre) ? maxL : (s.size() - pre);
            return maxL;
    }
};

int main()
{
    Solution test = Solution();
    string s = "qopubjguxhxdipfzwswybgfylqvjzhar";
    cout<<test.lengthOfLongestSubstring(s)<<endl;


    return 0;
}
